llmcam/store.py

```diff
@@ -141,111 +94,83 @@ def add_function_tools(
+
+# %% ../nbs/10_store.ipynb 31
+def execute_handler_core(
+    session_tools: dict, # Tools for each session
+    function_name: str,  # Name of the function to execute
+    module: str,  # Module of the function
+    session_id: str,  # Session ID
+    **kwargs  # Keyword arguments
+):
+    """Execute the handler function by retrieving function with session ID."""
 
-# %% ../nbs/10_store.ipynb 33
-def save_toolbox(destination: str = "toolbox.json"):
-    """Save toolbox to file."""
-    with open(destination, "w") as f:
-        json.dump(ToolBox, f)
+    # Get the toolbox for the session
+    tools = session_tools.get(session_id, None)
+    if tools is None:
+        raise ValueError(f"Session tools not found: {session_id}")
     
-#| export
-def load_toolbox(destination: str = "toolbox.json"):
-    """Load toolbox from file."""
-    global ToolBox
-    with open(destination, "r") as f:
-        ToolBox = json.load(f)
-
-# %% ../nbs/10_store.ipynb 34
-def extract_tools_from_services(
-    services: list[str] = []  # List of service names
-) -> list:  # List of tools
-    """Extract tools from services."""
-    # Initialize tools
-    tools = []
-
-    # Extract tools from services
-    global ToolBox
-    for service in services:
-        if service in ToolBox:
-            tools.extend(ToolBox[service])
-
-    # Raise error if no tools found
-    if len(tools) == 0:
-        raise ValueError("No tools found")
+    # Get the function
+    module = importlib.import_module(module)
+    function = getattr(module, function_name, None)
+    if function is None:
+        raise ValueError(f"Function not found: {function_name}")
     
-    # Raise error if too many tools
-    if len(tools) > 128:
-        raise ValueError("Too many tools for using GPT-4. Maximum number of tools is 128.")
-
-    # Return tools
-    return tools
-
-# %% ../nbs/10_store.ipynb 37
-add_api_tools(
-    service_name="road_digitraffic",
-    base_url="https://tie.digitraffic.fi",
-    oas_url="https://tie.digitraffic.fi/swagger/openapi.json",
-    oas_destination="api/road_digitraffic.json"
-)
-
-# %% ../nbs/10_store.ipynb 38
-add_api_tools(
-    service_name="train_digitraffic",
-    base_url="https://rata.digitraffic.fi",
-    oas_url="https://rata.digitraffic.fi/swagger/openapi.json",
-    oas_destination="api/train_digitraffic.json"
-)
-
-# %% ../nbs/10_store.ipynb 40
-add_function_tools(
-    service_name="ytube_live",
-    function_names=[
-        "llmcam.fn_to_fc.capture_youtube_live_frame_and_save",
-        "llmcam.fn_to_fc.ask_gpt4v_about_image_file"
-    ]
-)
-
-# %% ../nbs/10_store.ipynb 42
-from .oas_to_requests import generate_request
-
-def redirect_tool_calls(
-    function_name: str,  # Name of the function
-    tools: list = [],  # List of tools
-    **kwargs  # Keyword arguments
+    if "service" in kwargs:
+        del kwargs["service"]
+    
+    # Execute the function
+    function(tools, **kwargs)
+
+# %% ../nbs/10_store.ipynb 32
+def handler_schema(
```

Comment:
handler_* should be implemented ina class since FC won't call hander by fastHTML does?